File: tests/test_dynamic_array.py
import pytest

from data_structures.dynamic_array import DynamicArray


def test_append_and_index():
    arr = DynamicArray()
    for i in range(20):
        arr.append(i)
    assert len(arr) == 20
    assert arr.to_list() == list(range(20))


def test_resize_triggers_correctly():
    arr = DynamicArray(capacity=2)
    for i in range(10):
        arr.append(i)
    assert arr.to_list() == list(range(10))


def test_pop():
    arr = DynamicArray()
    for i in range(5):
        arr.append(i)
    assert arr.pop() == 4
    assert arr.pop() == 3
    assert len(arr) == 3


def test_pop_empty_raises():
    arr = DynamicArray()
    with pytest.raises(IndexError):
        arr.pop()


def test_insert_and_delete_middle():
    arr = DynamicArray()
    for i in [1, 2, 4, 5]:
        arr.append(i)
    arr.insert(2, 3)
    assert arr.to_list() == [1, 2, 3, 4, 5]
    deleted = arr.delete(0)
    assert deleted == 1
    assert arr.to_list() == [2, 3, 4, 5]


def test_get_set_bounds():
    arr = DynamicArray()
    arr.append(10)
    arr[0] = 99
    assert arr[0] == 99
    with pytest.raises(IndexError):
        arr[5]
    with pytest.raises(IndexError):
        arr[-1]

File: tests/test_dynamic_array.cpp
#include "../include/data_structures/dynamic_array.h"
#include "../include/test_framework.h"

TEST_CASE("dynamic array: append and get") {
    DynamicArray<int> arr;
    arr.append(10);
    arr.append(20);
    arr.append(30);
    CHECK(arr.size() == 3);
    CHECK(arr.get(0) == 10);
    CHECK(arr.get(1) == 20);
    CHECK(arr.get(2) == 30);
}

TEST_CASE("dynamic array: append triggers growth past initial capacity") {
    DynamicArray<int> arr;
    std::size_t initial_capacity = arr.capacity();
    for (int i = 0; i < 100; ++i) arr.append(i);
    CHECK(arr.size() == 100);
    CHECK(arr.capacity() > initial_capacity);
    for (int i = 0; i < 100; ++i) CHECK(arr.get(static_cast<std::size_t>(i)) == i);
}

TEST_CASE("dynamic array: insert in the middle shifts later elements") {
    DynamicArray<int> arr;
    arr.append(1);
    arr.append(3);
    arr.insert(1, 2);
    CHECK(arr.size() == 3);
    CHECK(arr.get(0) == 1);
    CHECK(arr.get(1) == 2);
    CHECK(arr.get(2) == 3);
}

TEST_CASE("dynamic array: remove_at shifts later elements left") {
    DynamicArray<int> arr;
    arr.append(1);
    arr.append(2);
    arr.append(3);
    arr.remove_at(1);
    CHECK(arr.size() == 2);
    CHECK(arr.get(0) == 1);
    CHECK(arr.get(1) == 3);
}

TEST_CASE("dynamic array: get/insert/remove out of range throws") {
    DynamicArray<int> arr;
    arr.append(1);
    bool threw = false;
    try {
        arr.get(5);
    } catch (const std::out_of_range&) {
        threw = true;
    }
    CHECK(threw);

    threw = false;
    try {
        arr.remove_at(5);
    } catch (const std::out_of_range&) {
        threw = true;
    }
    CHECK(threw);
}

TEST_CASE("dynamic array: empty array has size 0") {
    DynamicArray<int> arr;
    CHECK(arr.empty());
    CHECK(arr.size() == 0);
}

TEST_CASE("dynamic array: set overwrites existing element") {
    DynamicArray<int> arr;
    arr.append(1);
    arr.set(0, 99);
    CHECK(arr.get(0) == 99);
}
